File: node/Assembler/src/templateLoader/loaderNormal.js
// Normal template loader for Node.js - handles loading and caching of HTML templates from filesystem

import fs from 'fs';
import path from 'path';

export class TemplateResult {
    constructor(html, json = null) {
        this.html = html;
        this.json = json;
    }
}

export class LoaderNormal {
    static #htmlTemplatesCache = new Map();

    /**
     * Loads HTML files and corresponding JSON files from the specified application site directory, caching the output per appSite
     * @param {string} rootDirPath - Root directory path
     * @param {string} appSite - Application site name
     * @returns {Map<string, TemplateResult>} Map of templates
     */
    static loadGetTemplateFiles(rootDirPath, appSite) {
        const cacheKey = `${path.dirname(rootDirPath)}|${appSite}`;
        
        if (this.#htmlTemplatesCache.has(cacheKey)) {
            return this.#htmlTemplatesCache.get(cacheKey);
        }

        const result = new Map();
        const appSitesPath = path.join(rootDirPath, 'AppSites', appSite);
        
        if (!fs.existsSync(appSitesPath) || !fs.statSync(appSitesPath).isDirectory()) {
            this.#htmlTemplatesCache.set(cacheKey, result);
            return result;
        }

        // Recursively find all HTML files
        this.#walkDirectory(appSitesPath, (filePath, stats) => {
            if (stats.isFile() && path.extname(filePath) === '.html') {
                const fileName = path.basename(filePath, '.html');
                const key = `${appSite.toLowerCase()}_${fileName.toLowerCase()}`;
                
                const htmlContent = fs.readFileSync(filePath, 'utf8');
                const jsonFile = filePath.replace('.html', '.json');
                let jsonContent = null;
                
                if (fs.existsSync(jsonFile)) {
                    jsonContent = fs.readFileSync(jsonFile, 'utf8');
                }
                
                result.set(key, new TemplateResult(htmlContent, jsonContent));
            }
        });

        this.#htmlTemplatesCache.set(cacheKey, result);
        return result;
    }

    /**
     * Clear all cached templates (useful for testing or when templates change)
     */
    static clearCache() {
        this.#htmlTemplatesCache.clear();
    }

    /**
     * Helper method to recursively walk directory
     * @param {string} dir - Directory to walk
     * @param {Function} callback - Callback function for each file/directory
     */
    static #walkDirectory(dir, callback) {
        const files = fs.readdirSync(dir);
        
        for (const file of files) {
            const filePath = path.join(dir, file);
            const stats = fs.statSync(filePath);
            
            callback(filePath, stats);
            
            if (stats.isDirectory()) {
                this.#walkDirectory(filePath, callback);
            }
        }
    }
}

File: go/Assembler/loader/loader_normal.go
package loader

import (
	Logger "arshu/common"
	"assembler/common"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"strings"
	"sync"
)

var htmlTemplatesCache = struct {
	sync.RWMutex
	cache map[string]map[string]struct {
		HTML string
		JSON *string
	}
}{cache: make(map[string]map[string]struct {
	HTML string
	JSON *string
})}

// LoadGetTemplateFiles loads HTML and corresponding JSON files from the specified appSite directory, caching the output per appSite
// Returns map[string]struct{HTML string; JSON *string} equivalent to C#'s (string, string?) and Rust's (String, Option<String>)
func LoadGetTemplateFiles(rootDirPath, appSite, searchAppSites string) map[string]struct {
	HTML string
	JSON *string
} {
	Logger.Debug(fmt.Sprintf("LoadGetTemplateFiles called for appSite: %s, searchAppSites: %s", appSite, searchAppSites), "LoaderNormal")

	cacheKey := fmt.Sprintf("%s|%s|%s", filepath.Dir(rootDirPath), appSite, searchAppSites)

	htmlTemplatesCache.RLock()
	cached, ok := htmlTemplatesCache.cache[cacheKey]
	htmlTemplatesCache.RUnlock()
	if ok {
		Logger.Debug(fmt.Sprintf("Returning cached templates for %s (%d templates)", appSite, len(cached)), "LoaderNormal")
		return cached
	}

	// Load templates from primary appSite
	result := loadTemplatesFromSingleAppSiteNormal(rootDirPath, appSite)

	// Load templates from searchAppSites for fallback
	if searchAppSites != "" {
		searchAppSitesArray := strings.Split(searchAppSites, ",")
		for _, searchAppSite := range searchAppSitesArray {
			searchAppSite = strings.TrimSpace(searchAppSite)
			if searchAppSite == "" {
				continue
			}
			searchTemplates := loadTemplatesFromSingleAppSiteNormal(rootDirPath, searchAppSite)
			for k, v := range searchTemplates {
				if _, exists := result[k]; !exists {
					result[k] = v
					Logger.Debug(fmt.Sprintf("Added fallback template '%s' from '%s'", k, searchAppSite), "LoaderNormal")
				}
			}
		}
	}

	htmlTemplatesCache.Lock()
	htmlTemplatesCache.cache[cacheKey] = result
	htmlTemplatesCache.Unlock()
	return result
}

// loadTemplatesFromSingleAppSiteNormal loads templates from a single AppSite without caching or fallback logic
func loadTemplatesFromSingleAppSiteNormal(rootDirPath, appSite string) map[string]struct {
	HTML string
	JSON *string
} {
	result := make(map[string]struct {
		HTML string
		JSON *string
	})
	appSitesPath := filepath.Join(rootDirPath, "AppSites", appSite)
	if stat, err := os.Stat(appSitesPath); err != nil || !stat.IsDir() {
		Logger.Warn(fmt.Sprintf("AppSites directory not found: %s", appSitesPath), "LoaderNormal")
		return result
	}

	Logger.Debug(fmt.Sprintf("Loading templates from: %s", appSitesPath), "LoaderNormal")

	_ = filepath.Walk(appSitesPath, func(path string, info os.FileInfo, err error) error {
		if err != nil || info.IsDir() || !strings.HasSuffix(info.Name(), ".html") {
			return nil
		}
		fileName := strings.TrimSuffix(info.Name(), ".html")
		key := fmt.Sprintf("%s_%s", strings.ToLower(appSite), strings.ToLower(fileName))
		f, _ := os.Open(path)
		htmlBytes, _ := io.ReadAll(f)
		f.Close()
		htmlContent := common.NormalizeFileContent(string(htmlBytes))

		Logger.Debug(fmt.Sprintf("Loading template: %s (html size: %d)", key, len(htmlContent)), "LoaderNormal")

		// Find JSON file case-insensitively
		jsonFile := strings.TrimSuffix(path, ".html") + ".json"
		var jsonContent *string
		// Try exact match first
		if _, err := os.Stat(jsonFile); err == nil {
			jf, _ := os.Open(jsonFile)
			jsonBytes, _ := io.ReadAll(jf)
			jf.Close()
			jsonStr := common.NormalizeFileContent(string(jsonBytes))
			jsonContent = &jsonStr
			Logger.Debug(fmt.Sprintf("Found JSON file for %s (size: %d)", key, len(jsonStr)), "LoaderNormal")
		} else {
			// Try case-insensitive search in the same directory
			dir := filepath.Dir(path)
			baseName := strings.ToLower(strings.TrimSuffix(filepath.Base(path), ".html"))
			entries, err := os.ReadDir(dir)
			if err == nil {
				for _, entry := range entries {
					if !entry.IsDir() && strings.HasSuffix(strings.ToLower(entry.Name()), ".json") {
						entryBase := strings.ToLower(strings.TrimSuffix(entry.Name(), ".json"))
						if entryBase == baseName {
							matchedJsonPath := filepath.Join(dir, entry.Name())
							jf, _ := os.Open(matchedJsonPath)
							jsonBytes, _ := io.ReadAll(jf)
							jf.Close()
							jsonStr := common.NormalizeFileContent(string(jsonBytes))
							jsonContent = &jsonStr
							Logger.Debug(fmt.Sprintf("Found JSON file (case-insensitive) for %s (size: %d)", key, len(jsonStr)), "LoaderNormal")
							break
						}
					}
				}
			}
		}
		result[key] = struct {
			HTML string
			JSON *string
		}{HTML: htmlContent, JSON: jsonContent}
		return nil
	})

	Logger.Debug(fmt.Sprintf("Loaded %d templates for %s", len(result), appSite), "LoaderNormal")
	return result
}

// ClearNormalCache clears all cached normal templates (useful for testing or when templates change)
func ClearNormalCache() {
	htmlTemplatesCache.Lock()
	htmlTemplatesCache.cache = make(map[string]map[string]struct {
		HTML string
		JSON *string
	})
	htmlTemplatesCache.Unlock()
}
